_examples/oauth2.go

```diff
@@ -0,0 +1,364 @@
+			usernameConfig,
+			passwordConfig,
+		)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		w.Header().Set("Content-Type", "application/json")
+		globalToken = token
+		e := json.NewEncoder(w)
+		e.SetIndent("", "  ")
+		_ = e.Encode(token)
+	})
+
+	mux.HandleFunc("/client", func(w http.ResponseWriter, r *http.Request) {
+		cfg := clientcredentials.Config{
+			ClientID:     authConfig.ClientID,
+			ClientSecret: authConfig.ClientSecret,
+			TokenURL:     authConfig.Endpoint.TokenURL,
+		}
+
+		token, err := cfg.Token(context.Background())
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		w.Header().Set("Content-Type", "application/json")
+		e := json.NewEncoder(w)
+		e.SetIndent("", "  ")
+		_ = e.Encode(token)
+	})
+
+	return mux
+}
+
+func dumpRequest(writer io.Writer, header string, r *http.Request) error {
+	data, err := httputil.DumpRequest(r, true)
+	if err != nil {
+		return err
+	}
+	_, _ = writer.Write([]byte("\n" + header + ": \n"))
+	_, _ = writer.Write(data)
+	return nil
+}
+
+func userAuthorizeHandler(
+	w http.ResponseWriter,
+	r *http.Request,
+) (userID string, err error) {
+	_ = dumpRequest(os.Stdout, "userAuthorizeHandler", r)
+
+	st, err := session.Start(r.Context(), w, r)
+	if err != nil {
+		return
+	}
+
+	uid, ok := st.Get("LoggedInUserID")
+	if !ok {
+		if r.Form == nil {
+			_ = r.ParseForm()
+		}
+
+		st.Set("ReturnUri", r.Form)
+		_ = st.Save()
+
+		w.Header().Set("Location", "/login")
+		w.WriteHeader(http.StatusFound)
+		return
+	}
+
+	userID = uid.(string)
+	st.Delete("LoggedInUserID")
+	_ = st.Save()
+	return
+}
+
+func loginHandler(w http.ResponseWriter, r *http.Request) {
+	_ = dumpRequest(os.Stdout, "login", r)
+
+	st, err := session.Start(r.Context(), w, r)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	if r.Method == "POST" {
+		if r.Form == nil {
+			if err := r.ParseForm(); err != nil {
+				http.Error(w, err.Error(), http.StatusInternalServerError)
+				return
+			}
+		}
+
+		if r.Form.Get("username") != usernameConfig &&
+			r.Form.Get("password") != passwordConfig {
+			w.WriteHeader(http.StatusBadRequest)
+			return
+		}
+
+		st.Set("LoggedInUserID", r.Form.Get("username"))
+		_ = st.Save()
+
+		w.Header().Set("Location", "/auth")
+		w.WriteHeader(http.StatusFound)
+		return
+	}
+
+	outputHTML(w, r, "static/login.html")
+}
+
+func authHandler(w http.ResponseWriter, r *http.Request) {
+	_ = dumpRequest(os.Stdout, "auth", r)
+
+	st, err := session.Start(r.Context(), w, r)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	if _, ok := st.Get("LoggedInUserID"); !ok {
+		w.Header().Set("Location", "/login")
+		w.WriteHeader(http.StatusFound)
+		return
+	}
+
+	outputHTML(w, r, "static/auth.html")
+}
+
+func outputHTML(w http.ResponseWriter, req *http.Request, filename string) {
```

Comment:
Why do we need html pages for this example? Could we remove them?

Reply:
Noted, I'll remove unnecessary things here.

Reply:
Addressed